ts-web: add browser extension utilities

pro-wh opened the pull request:
summary

a browser extension "signer interface." where a browser extension can serve the role of a ContextSigner implementation.

---

update zero

Turns out there's a feature `web_accessible_resources`, which will allow us to make a webpage that sites can load into an iframe and then message. I've written up some code to verify that this works.

Next I'll be writing some helper functions to interact with the extension within the `ContextSigner` interface.

pro-wh commented:
I've extracted some code into a library. It's missing some things we usually expect--tests, package name. But it's at a point where reviewers can take a look to get an impression of the overall design

kostko commented on client-sdk/ts-web/ext-utils/src/protocol.ts:
```diff
@@ -0,0 +1,69 @@
+    method: typeof METHOD_CONTEXT_SIGNER_PUBLIC;
+    /**
+     * An extra parameter, suggested to be used to help select which key to
+     * use. This lirbary passes it through verbatim; it's up to the extension
```
```suggestion
     * use. This library passes it through verbatim; it's up to the extension
```

kostko commented on client-sdk/ts-web/ext-utils/src/protocol.ts:
```diff
@@ -0,0 +1,69 @@
+     * use. This lirbary passes it through verbatim; it's up to the extension
+     * to define how it's interpreted.
+     */
+    which: any;
```
Could be useful to have a method for listing keys as well?

pro-wh replied:
o good idea

pro-wh replied:
ok I tried adding one

kostko replied:
Cool, looks good, thanks!

kostko commented on client-sdk/ts-web/ext-utils/src/protocol.ts:
```diff
@@ -0,0 +1,69 @@
+    method: typeof METHOD_CONTEXT_SIGNER_SIGN;
+    /**
+     * An extra parameter, suggested to be used to help select which key to
+     * use. This lirbary passes it through verbatim; it's up to the extension
```
```suggestion
     * use. This library passes it through verbatim; it's up to the extension
```

kostko reviewed (commented):
Cool, I like the general idea of how this works. For the API is there anything else that might be useful to have? Also should there be support for versioning the protocol?

pro-wh commented:
I tried adding tests but the headless version can't click the `window.confirm`s :laughing:

pro-wh commented:
could maybe do some versioning via method names. maybe we'd like to version the overall transport where "it makes an iframe and postmessage with type=MESSAGE_REQUEST etc" is version 1 as well

kostko commented on client-sdk/ts-web/ext-utils/src/protocol.ts:
```diff
@@ -0,0 +1,92 @@
+
+// methods
+
+export const METHOD_KEYS_LIST = 'list-keys';
```
```suggestion
export const METHOD_KEYS_LIST = 'keys-list';
```
Either that or rename all the types/vars.

pro-wh commented:
I added some logic to make it not open any alerts and confirms under a special testing mode. Under this mode it also uses a fixed key from the ADR 0008 test vectors.

I added some `-v1` suffixes to message/method constants.

kostko reviewed (approved):
I see this is shared with #233.

pro-wh commented:
correction to cypress screenshot upload step. gonna merge
